File: packages/sdk-py/src/agentgate/agent_gate.py
"""AgentGate — main entry point for connecting an agent to the platform."""
from __future__ import annotations
import asyncio
import logging
from typing import Callable, Awaitable, Optional

from .ws_client import WsClient
from .message_context import MessageContext
from .crypto import verify_payload_hmac

logger = logging.getLogger(__name__)

MessageHandler = Callable[[MessageContext], Awaitable[None]]


class AgentGate:
    """
    Connect an agent to AgentGate.

    Usage:
        gate = AgentGate(token="...", agent_id="...", gateway_url="wss://...")
        gate.on_message(handler)
        await gate.connect()
    """

    def __init__(
        self,
        token: str,
        agent_id: str,
        gateway_url: str,
        reconnect: bool = True,
        reconnect_max_delay: float = 30.0,
    ) -> None:
        self._token = token
        self._agent_id = agent_id
        url = f"{gateway_url}/ws/agent?agentId={agent_id}&token={token}"

        self._ws = WsClient(url, reconnect=reconnect, reconnect_max_delay=reconnect_max_delay)
        self._ws.on_message(self._dispatch)
        self._handler: Optional[MessageHandler] = None

    def on_message(self, handler: MessageHandler) -> "AgentGate":
        """Register the async handler for incoming messages."""
        self._handler = handler
        return self

    async def connect(self) -> None:
        """Start the WebSocket connection (blocks until disconnected)."""
        logger.info("[AgentGate] Connecting agent %s", self._agent_id)
        await self._ws.connect()

    def disconnect(self) -> None:
        """Stop reconnecting and close the connection."""
        self._ws.disconnect()

    async def _dispatch(self, msg: dict) -> None:
        """Called for every incoming frame. Verifies HMAC and calls handler."""
        if msg.get("type") != "message":
            return

        message_id = msg.get("messageId")
        payload = msg.get("payload")

        if not isinstance(payload, dict) or not message_id:
            logger.warning("[AgentGate] Malformed frame received")
            return

        if not verify_payload_hmac(payload, self._token):
            logger.warning("[AgentGate] Dropping message — HMAC verification failed.")
            return

        if not self._handler:
            logger.warning("[AgentGate] No handler registered. Call on_message() before connect().")
            return

        ctx = MessageContext(message_id, payload, self._ws)

        try:
            await self._handler(ctx)
        except Exception as exc:
            logger.error("[AgentGate] Handler raised an exception: %s", exc)

File: packages/sdk-ts/src/agent-gate.ts
import { WsClient } from './ws-client'
import { MessageContext } from './message-context'
import { verifyPayloadHmac } from './crypto'
import type { AgentGateOptions, IncomingFrame, UserContext } from './types'

type MessageHandler = (ctx: MessageContext) => Promise<void>

export class AgentGate {
  private readonly ws: WsClient
  private handler: MessageHandler | null = null

  constructor(private readonly options: AgentGateOptions) {
    const { agentId, token, gatewayUrl } = options
    const url = `${gatewayUrl}/ws/agent?agentId=${agentId}&token=${token}`

    this.ws = new WsClient({
      url,
      reconnect: options.reconnect ?? true,
      reconnectMaxDelay: options.reconnectMaxDelay ?? 30000,
    })

    this.ws.on('message', (msg: unknown) => void this.dispatch(msg))
    this.ws.on('open', () => console.log('[AgentGate] Connected'))
    this.ws.on('close', () => console.log('[AgentGate] Disconnected'))
    this.ws.on('error', (err: Error) => console.error('[AgentGate] Error:', err.message))
  }

  onMessage(handler: MessageHandler): this {
    this.handler = handler
    return this
  }

  connect(): this {
    this.ws.connect()
    return this
  }

  disconnect(): void {
    this.ws.disconnect()
  }

  private async dispatch(msg: unknown): Promise<void> {
    if (!isIncomingFrame(msg)) return

    const { messageId, payload } = msg

    if (!verifyPayloadHmac(payload as unknown as Record<string, unknown>, this.options.token)) {
      console.warn('[AgentGate] Dropping message — HMAC verification failed.')
      return
    }

    if (!this.handler) {
      console.warn('[AgentGate] Received a message but no handler is registered.')
      return
    }

    const { hmac: _hmac, ...userContext } = payload
    const ctx = new MessageContext(messageId, userContext as UserContext, this.ws)

    try {
      await this.handler(ctx)
    } catch (err) {
      console.error('[AgentGate] Handler threw an error:', err)
    }
  }
}

function isIncomingFrame(msg: unknown): msg is IncomingFrame {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    (msg as IncomingFrame).type === 'message' &&
    typeof (msg as IncomingFrame).messageId === 'string' &&
    typeof (msg as IncomingFrame).payload === 'object' &&
    (msg as IncomingFrame).payload !== null
  )
}
